### PR TITLE
add local logging for debugging

### PR DESCRIPTION
**Pull request by tarkah**
First commit is the meat of adding logging support for Ajour. On startup, it will create / truncate the log file so it should never grow too big and is only relevant to the last launch of the app.

- If `cargo run` it will log to console.

- If `cargo run --release` / binary is used, it will log to `$config/ajour/ajour.log`.

The rest is just going to be scattering `debug!` everywhere! I figured we can roll this out with everything as `DEBUG` and if needed, convert some stuff to `INFO` or `TRACE` as needed if there is too much noise.

### CODE REVIEW

**Comment by tarkah** [comment]
This is going to be really nice for when we add new features, especially when modifying the `parse` module where I plan to put most of the detailed debugging.

Right now I've got just general flow logged from the `update` module:

```
   Compiling ajour v0.3.3 (C:\Users\cforsstrom\Documents\wsl\rust\ajour)
    Finished dev [unoptimized + debuginfo] target(s) in 4.93s
     Running `target\debug\ajour.exe`
13:16:03.930 [ajour][DEBUG] Ajour has started.
13:16:04.969 [ajour::config][DEBUG] loading config
13:16:04.969 [ajour::utility][DEBUG] checking for application update
13:16:04.969 [ajour::theme][DEBUG] loading user themes
13:16:04.973 [ajour::fs::theme][DEBUG] loaded 1 user themes
13:16:05.079 [ajour::gui::update][DEBUG] handling Message::Parse
13:16:05.079 [ajour::gui::update][DEBUG] handling Message::ThemesLoaded(1 themes)
13:16:05.348 [ajour::gui::update][DEBUG] handling Message::NeedsUpdate(Some("0.3.2"))
13:16:09.905 [ajour::gui::update][DEBUG] handling Message::ParsedAddons(16 addons)
13:16:58.238 [ajour::gui::update][DEBUG] handling Interaction::Settings
13:17:00.722 [ajour::gui::update][DEBUG] handling Message::FlavorSelected(retail)
13:17:00.726 [ajour::config][DEBUG] loading config
13:17:00.749 [ajour::gui::update][DEBUG] handling Message::Parse
13:17:00.780 [ajour::gui::update][ERROR] Addon directory not found: "C:\\Program Files (x86)\\World of Warcraft\\_retail_\\Interface/AddOns"
13:17:15.829 [ajour::gui::update][DEBUG] handling Message::FlavorSelected(classic)
13:17:15.831 [ajour::config][DEBUG] loading config
13:17:15.852 [ajour::gui::update][DEBUG] handling Message::Parse
13:17:16.698 [ajour::gui::update][DEBUG] handling Message::ParsedAddons(16 addons)
13:17:19.090 [ajour::gui::update][DEBUG] handling Message::ThemeSelected("Light")
13:17:20.720 [ajour::gui::update][DEBUG] handling Message::ThemeSelected("Dracula")
13:17:22.284 [ajour::gui::update][DEBUG] handling Message::ThemeSelected("Dark")
13:17:24.781 [ajour::gui::update][DEBUG] handling Interaction::Settings
13:17:33.317 [ajour::gui::update][DEBUG] handling Interaction::Update(ClassicCastbars)
13:17:33.935 [ajour::gui::update][DEBUG] handling Message::DownloadedAddon((ClassicCastbars, error: false))
13:17:34.122 [ajour::gui::update][DEBUG] handling Message::UnpackedAddon((ClassicCastbars, error: false))
13:17:34.158 [ajour::gui::update][DEBUG] handling Message::UpdateFingerprint((ClassicCastbars, error: false))
13:17:34.237 [ajour::gui::update][DEBUG] handling Message::UpdateFingerprint((ClassicCastbars_Options, error: false))
13:17:42.198 [ajour::gui::update][DEBUG] handling Interaction::Refresh
13:17:42.202 [ajour::config][DEBUG] loading config
13:17:42.214 [ajour::gui::update][DEBUG] handling Message::Parse
13:17:43.239 [ajour::gui::update][DEBUG] handling Message::ParsedAddons(16 addons)
```

**Comment by casperstorm** [comment]
That is REALLY nice. We could maybe even drop `handling` from the `Message` and `Interaction` calls?

**Comment by tarkah** [comment]
Good call, I should base it off #108 so I don't have a nasty merge 👍

**Comment by casperstorm** [comment]
> Good call, I should base it off #108 so I don't have a nasty merge 👍

Yeah, and #108 is soon done – just some logic around loading state etc, but shouldn't be that much so good idea to branch from it if possible.

**Comment by tarkah** [comment]
Great, I just rebased to that branch and removed `handling `

**Comment by casperstorm** [comment]
@tarkah do you switch it to ready for review when you are ready?

**Comment by tarkah** [comment]
Yeah I'll switch it over. I was going to hold off on doing anything else until we merged #108. Most of the lines I've added are very high level, I think it'd make sense to add some TRACE debugging that goes into some of the per addon parsing / fingerprinting logic as well. Would be interested for you to checkout this branch, `cargo run` and give me an idea on what else you'd like to see for debugging.

**Comment by casperstorm** [comment]
> Yeah I'll switch it over when ready. I was going to hold off on doing anything else until we merged #108. Most of the lines I've added are very high level, I think it'd make sense to add some TRACE debugging that goes into some of the per addon parsing / fingerprinting logic as well. Would be interested for you to checkout this branch, `cargo run` and give me an idea on what else you'd like to see for debugging.

I'll make a list of what we should have once I wake up tomorrow.

**Comment by casperstorm** [comment]
- [src/parse.rs](https://github.com/casperstorm/ajour/blob/73424bf3ce37d2c50732f7bd8d86c012ea238c1e/src/parse.rs#L291). Log amount of total hashes `fingerprint_hashes.len()`
- [src/parse.rs](https://github.com/casperstorm/ajour/blob/73424bf3ce37d2c50732f7bd8d86c012ea238c1e/src/parse.rs#L299). I think its more relevant to show  `fingerprint_package.exact_matches.len()` rather than `fingerprint_package.exact_fingerprints.len()` since each `excat_fingerprint` is what we convert to a visible addon.
- [src/gui/update.rs](https://github.com/casperstorm/ajour/blob/73424bf3ce37d2c50732f7bd8d86c012ea238c1e/src/gui/update.rs#L57). We should log when we enter here because that means they didn't have a wow path.
- [src/gui/update.rs](https://github.com/casperstorm/ajour/blob/73424bf3ce37d2c50732f7bd8d86c012ea238c1e/src/gui/update.rs#L48). Here we can log the wow path which we are about to parse.
- [src/gui/update.rs](https://github.com/casperstorm/ajour/blob/73424bf3ce37d2c50732f7bd8d86c012ea238c1e/src/gui/update.rs#L33). Should we consider actually logging the parsed config? 🤔

**Comment by tarkah** [comment]
@casperstorm I've added all your suggestions. However, to clarify this one:


>     * If hashing fails, it could be relevant to know the error, and which addon caused the failure.

Do you mean the computation of the hash, or the lookup against the API? I added trace logging on partial matches against the API.

Checkout how this looks below:

```
10:35:17.831 [ajour][DEBUG] Ajour 0.3.3 has started.
10:35:18.969 [ajour::config][DEBUG] loading config
10:35:18.969 [ajour::utility][DEBUG] checking for application update
10:35:18.969 [ajour::theme][DEBUG] loading user themes
10:35:18.974 [ajour::fs::theme][DEBUG] loaded 2 user themes
10:35:19.106 [ajour::gui::update][DEBUG] Message::Parse
10:35:19.106 [ajour::gui::update][DEBUG] config loaded:
Config {
    wow: Wow {
        directory: Some(
            "C:\\Program Files (x86)\\World of Warcraft",
        ),
        flavor: Classic,
    },
    addons: Addons {
        ignored: {
            Classic: [],
            Retail: [],
        },
    },
    theme: None,
}
10:35:19.141 [ajour::gui::update][DEBUG] preparing to parse addons in "C:\\Program Files (x86)\\World of Warcraft\\_retail_\\Interface/AddOns"
10:35:19.143 [ajour::gui::update][DEBUG] preparing to parse addons in "C:\\Program Files (x86)\\World of Warcraft\\_classic_\\Interface/AddOns"
10:35:19.155 [ajour::parse][DEBUG] classic - parsing addons folder
10:35:19.154 [ajour::parse][DEBUG] retail - parsing addons folder
10:35:19.158 [ajour::parse][DEBUG] retail - 22 folders in AddOns directory to parse
10:35:19.343 [ajour::gui::update][DEBUG] Message::NeedsUpdate(Some("0.3.2"))
10:35:19.383 [ajour::parse][DEBUG] classic - 67 fingerprints: 67 cached, 0 calculated, 0 added, 0 removed
10:35:19.454 [ajour::parse][DEBUG] classic - 67 successfully parsed from '.toc'
10:35:19.456 [ajour::parse][DEBUG] classic - 2 addons with tukui id
10:35:19.456 [ajour::parse][DEBUG] retail - 22 fingerprints: 22 cached, 0 calculated, 0 added, 0 removed
10:35:19.494 [ajour::parse][DEBUG] retail - 22 successfully parsed from '.toc'
10:35:19.502 [ajour::parse][DEBUG] retail - 0 addons with tukui id
10:35:19.502 [ajour::parse][DEBUG] retail - 0 addons applied with tukui id package metadata
10:35:19.503 [ajour::parse][DEBUG] retail - 22 unique fingerprints to check against curse api
10:35:19.738 [ajour::parse][TRACE] retail - partial fingerprint found:
        Curse ID: 3025116
        File Name: DBM-DragonSoul-r209
        Modules (Name - Local Fingerprint - Remote Fingerprint)
        DBM-DragonSoul - 2190768360 - 353973115
        DBM-BaradinHold - 2298533366 - 2298533366
        DBM-BastionTwilight - 1291395356 - 1291395356
        DBM-BlackwingDescent - 2892435859 - 2892435859
        DBM-Firelands - 3733420408 - 3733420408
        DBM-ThroneFourWinds - 614383502 - 614383502
10:35:19.758 [ajour::parse][DEBUG] retail - 1 exact fingerprint matches against curse api
10:35:19.769 [ajour::parse][DEBUG] retail - 1 addons applied with fingerprint metadata
10:35:19.770 [ajour::parse][DEBUG] retail - 1 addons with curse id
10:35:20.033 [ajour::parse][DEBUG] retail - 1 addons applied with curse id package metadata
10:35:20.034 [ajour::parse][DEBUG] retail - 1 addons successfully parsed
10:35:20.036 [ajour::gui::update][DEBUG] Message::ParsedAddons(retail, 1 addons)
10:35:20.271 [ajour::parse][DEBUG] classic - 2 addons applied with tukui id package metadata
10:35:20.272 [ajour::parse][DEBUG] classic - 65 unique fingerprints to check against curse api
10:35:20.748 [ajour::parse][TRACE] classic - partial fingerprint found:
        Curse ID: 3050298
        File Name: ClassicCastbars-1.3.12-classic
        Modules (Name - Local Fingerprint - Remote Fingerprint)
        ClassicCastbars_Options - 2861130357 - 3434670468
        ClassicCastbars - 958824653 - 958824653
10:35:20.765 [ajour::parse][DEBUG] classic - 15 exact fingerprint matches against curse api
10:35:20.765 [ajour::parse][DEBUG] classic - 15 addons applied with fingerprint metadata
10:35:20.767 [ajour::parse][DEBUG] classic - 15 addons with curse id
10:35:21.369 [ajour::parse][DEBUG] classic - 15 addons applied with curse id package metadata
10:35:21.370 [ajour::parse][DEBUG] classic - 17 addons successfully parsed
10:35:21.375 [ajour::gui::update][DEBUG] Message::ParsedAddons(classic, 17 addons)
```

**Comment by casperstorm** [comment]
@tarkah Good morning ☕️
Nice, I think it's nice to have the above mentioned so super good job.

> Do you mean the computation of the hash, or the lookup against the API? I added trace logging on partial matches against the API.

I was actually thinking if this fails:
https://github.com/casperstorm/ajour/blob/868937b26b3c2819b4dded3b37470e1d9989fa4f/src/parse.rs#L372-L377

If SOMETHING goes wrong here it might be nice to know because then we won't get a fingerprint for that addon.

**Comment by tarkah** [comment]
Ahh yes, we need to convert that function to return a Result and identify which operations it can fail on and give them decent error messages. Then we can log the error if it returns one.

**Comment by tarkah** [comment]
Ok, check out the latest push. I've mapped any potential error to `ClientError::Fingerprint` and then log that as an `ERROR` if it fails

File: src/parse.rs
Comment (tarkah) [inline src/parse.rs:644]: We can't break from an iterator using `?` so I converted this to a for loop

**Review by casperstorm — APPROVED** [review]
This is a really really good addition. I'll merge it to master now.
Then we can test master and see what state we are in!

**Comment by tarkah** [comment]
Awesome!! Can't wait for the next release :)

**Comment by casperstorm** [comment]
@tarkah  - Hmm, I just tried to compile now on my macOS and i get `usefofeunstable library feature 'str_strip'`.

**Comment by tarkah** [comment]
`rustup update` all the time!! :P

**Pull request merged**
